sharding/review/casper_v2.1.md

```diff
@@ -0,0 +1,380 @@
+
+For a block on the beacon chain to be processed by a node, three conditions have to be met:
+
+* The parent pointed to by the `parent_hash` has already been processed and accepted
+* The PoW chain block pointed to by the `pow_chain_ref` has already been processed and accepted
+* The node's local clock time is greater than or equal to the minimum timestamp as computed by `GENESIS_TIME + slot_number * SLOT_DURATION`
+
+If these three conditions are not met, the client should delay processing the block until the three conditions are all satisfied.
+
+Block production is significantly different because of the proof of stake mechanism. A client simply checks what it thinks is the canonical chain when it should create a block, and looks up what its slot number is; when the slot arrives, it either proposes or attests to a block as required.
+
+### Beacon chain fork choice rule
+
+The beacon chain uses the Casper FFG fork choice rule of "favor the chain containing the highest-epoch justified checkpoint". To choose between chains that are all descended from the same justified checkpoint, the chain uses "recursive proximity to justification" (RPJ) to choose a checkpoint, then uses GHOST within an epoch.
+
+RPJ works as follows. First, start at the latest justified checkpoint. Then, choose the descendant checkpoint that has the most attestations supporting it. Repeat until you get to a checkpoint with no children (in this phase, only checkpoints whose epochs are completed, ie. epoch N where the timeslots for epoch N+1 have already begun, are admissible)
+
+For example:
+
+![](https://vitalik.ca/files/8d100380.png =225x600)
+
+From the child chosen by RPJ, use GHOST to find the head of the chain.
+
+## Beacon chain state transition function
+
+We now define the state transition function. At the high level, the state transition is made up of two parts:
+
+1. The epoch transition, which happens only if `slot_number // EPOCH_LENGTH > epoch_number`, and affects the `CrystallizedState` and `ActiveState`
+2. The per-block processing, which happens every block (if during an epoch transition block, it happens after the epoch transition), and affects the `ActiveState` only
+
+The epoch transition generally focuses on changes to the validator set, including adjusting balances and adding and removing validators, as well as processing crosslinks and setting up FFG checkpoints, and the per-block processing generally focuses on verifying aggregate signatures and saving temporary records relating to the in-block activity in the `ActiveState`.
+
+### Helper functions
+
+We start off by defining some helper algorithms. First, an algorithm for pseudorandomly shuffling the validator set based on some seed:
+
+```python
+def get_shuffling(seed, validator_count):
+    assert validator_count <= 16777216
```

Comment:
Why `16777216` (2^24) while `MAX_VALIDATOR_COUNT` is defined as 2^22 = 4194304?

Reply:
The algorithm very directly depends on the number of validators being less than 2^24 because of how it gets indices three bytes at a time.

Reply:
Got it 👍
Maybe use `MAX_VALIDATOR_COUNT << 2`?